capactio/capact
# Implement Secret storage backend as gRPC server

## Description

Changes proposed in this pull request:

- Add Protocol Buffers schema
- Generate gRPC client + server code from Protocol Buffers
- Implement gRPC server for secret storage backend
    - Support initially just AWS Secrets Manager. Tested also `dotenv` provider. The `vault` provider needs to be tested
    - Add tests for the gRPC server with fake secret provider 

## TODO

- [x] Update Development guide on website: https://github.com/capactio/website/pull/110
- [x] Create PR with dotenv and AWS Secrets Manager for teller repo (https://github.com/SpectralOps/teller/pull/60)

## Notes

Lint failing because of timeout. Don't worry about it - The job will be green after review 🙂 

## Testing

To regenerate the gRPC code, run:

```bash
SKIP_DEPS_INSTALLATION=false make gen-grpc-resources
```

To run example calls against the gRPC service:

1. Create AWS security credentials with `SecretsManagerReadWrite` policy.
2. Export environment variables:
    
   ```bash
    export AWS_ACCESS_KEY_ID="{accessKey}"
    export AWS_SECRET_ACCESS_KEY="{secretKey}"
    ```

3. Run the server:

    ```bash
    APP_SUPPORTED_PROVIDERS=dotenv,aws_secretsmanager APP_LOGGER_DEV_MODE=true go run ./cmd/secret-storage-backend/main.go
    ```

4. Now let's use Go generated gRPC client to to see how the server behaves. Run the example [`pkg/hub/api/grpc/storage_backend/example_test.go`](https://github.com/capactio/capact/blob/0f3c7b2b889192af54b9cd38dc45e26ce7315560/pkg/hub/api/grpc/storage_backend/example_test.go).

    To run it, you can e.g. change the `func ExampleNewStorageBackendClient() {` to `func TestE2EScenario(t *testing.T) {` and run it as an usual test.

6. Change the line 12 of this test and run it again:

    ```go
    provider := "aws_secretsmanager"
    ```


## Related issue(s)

Resolves #637
